### DISCUSSION TITLE
Blink.cmp completions doesn't seem to work? 🤔

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [olimorris/codecompanion.nvim#578] [Q&A] [answered]
I just recently noticed that completion for `#` and `@` commands doesn't work.

<img width="304" alt="image" src="https://github.com/user-attachments/assets/eeae9c65-f810-4843-a185-0dfee4c91e0e" />

@olimorris is it just me, or do you also not see any completions?

**Comment by olimorris** [comment]
It works for me. I was using it in the `V11.0.0` announcement video. Although it's outputting some brackets for some reason. The Blink API has changed so much in the last week. What's your config look like?

**Reply by fredrikaverpil** [reply]
That's good to hear, then it's likely just something on my end. I wonder what though...

- [blink.cmp config](https://github.com/fredrikaverpil/dotfiles/blob/98dc76c0516796963b5eb4babaa790374986f8ba/nvim-fredrik/lua/fredrik/plugins/core/blink.lua)
- [codecompanion.nvim config](https://github.com/fredrikaverpil/dotfiles/blob/98dc76c0516796963b5eb4babaa790374986f8ba/nvim-fredrik/lua/fredrik/plugins/codecompanion.lua)

**Reply by fredrikaverpil** [reply]
Ah, there's something wrong with how `sources.default` is extended. This is the final blink.cmp `opts` spec:

```lua
sources = {
  default = { "dadbod", "lsp", "path", "snippets", "buffer" },
  providers = {
    codecompanion = {
      enabled = true,
      module = "codecompanion.providers.completion.blink",
      name = "CodeCompanion"
    },
    dadbod = {
      module = "vim_dadbod_completion.blink",
      name = "Dadbod"
    },
    lazydev = {
      module = "lazydev.integrations.blink",
      name = "LazyDev",
      score_offset = 100
    },
    markdown = {
      module = "render-markdown.integ.blink",
      name = "RenderMarkdown"
    },
    obsidian = {
      module = "blink.compat.source",
      name = "obsidian"
    },
    obsidian_new = {
      module = "blink.compat.source",
      name = "obsidian_new"
    },
    obsidian_tags = {
      module = "blink.compat.source",
      name = "obsidian_tags"
    },
    snippets = {
      opts = {
        friendly_snippets = true,
        search_paths = { "/Users/fredrik/.dotfiles/nvim-fredrik/snippets" }
      }
    }
  }
```

Thanks for rubberducking 😉 then the error is on my end.

**Reply by fredrikaverpil** [reply, marked as answer]
I have multiple blink.cmp configs, and in each one, I was missing this:

```lua
{
    "saghen/blink.cmp",
    opts_extend = {
      "sources.default",
    },
}
```

**Reply by olimorris** [reply]
I've switched between nvim-cmp and blink so many times in the last two weeks. Blink in the cmdline was crashing Neovim. That got fixed. Then it was throwing a random error which needed the Rust dependencies rebuilding. It's been painful. I can't wait for a 1.0 release for it

**Reply by bassamsdata** [reply]
I’m not sure if this is related, but` codecompanion` completions don’t seem to work with a fresh installation of LazyVim. It could be an issue with `blink.cmp` or `fzf-lua`, as LazyVim now uses these by [default](https://www.lazyvim.org/news#added-plugins) instead of telescope and nvim-cmp. 

I don’t use LazyVim myself, but someone on Reddit mentioned that they couldn’t get any codecompanion completions to work in LazyVim.
and it might be related to issue #577

edit: removed some links

![Screen Recording 2024-12-30 at 10 37 20 AM](https://github.com/user-attachments/assets/a1723fa5-7faa-451f-a0fe-da3ce30faaee)

**Reply by fredrikaverpil** [reply]
> I don’t use LazyVim myself, but someone on Reddit mentioned that they couldn’t get any codecompanion completions to work in LazyVim.

@bassamsdata yes it looks like the same issue as I had, given your screencast above.

LazyVim does it right, and includes the following: https://github.com/LazyVim/LazyVim/blob/d0c366e4d861b848bdc710696d5311dca2c6d540/lua/lazyvim/plugins/extras/coding/blink.lua#L19-L23

However, when you add codecompanion (as it does not seem like it exists as an extra?), you have to include this `opts_extend` entry in your config too. I can see from your screencast above that you don't have a `opts_extend` key in the plugin spec. This is btw fairly new and a really awesome addition to the lazy.nvim plugin manager.


> and it might be related to issue https://github.com/olimorris/codecompanion.nvim/issues/577

Not sure about that though. In case it helps, I have codecompanion completions working in my config which closely resembles LazyVim: https://github.com/fredrikaverpil/dotfiles/tree/main/nvim-fredrik
Note that I'm experimenting with `vim.lsp.config` and `vim.lsp.enable` which requires Neovim nightly.

**Reply by olimorris** [reply]
> Note that I'm experimenting with `vim.lsp.config` and `vim.lsp.enable` which requires Neovim nightly.

I shall be perusing your dotfiles with great interest 😉. Have you checked out native snippets yet?

**Reply by fredrikaverpil** [reply]
> I shall be perusing your dotfiles with great interest

Haha, don't hesitate to reach out or let me know if you have questions!

> Have you checked out native snippets yet?

Yes, I'm using native snippets! 👍 
Look at my blink.cmp setup ([here](https://github.com/fredrikaverpil/dotfiles/blob/26f9bb08334e24835fb35584276b838cd5ee9d26/nvim-fredrik/lua/fredrik/plugins/core/blink.lua#L43)) for how I use friendly snippets.
Here's one: https://github.com/fredrikaverpil/dotfiles/blob/26f9bb08334e24835fb35584276b838cd5ee9d26/nvim-fredrik/snippets/lua.json

**Reply by bassamsdata** [reply]
@fredrikaverpil Thank you for the information. I added that and almost resolved the issue, but I overlooked that I need to change the provider to fzf_lua. I assumed that would be automatic if the telescope isn’t found.

However, even after making that change, the only thing that worked was using the `/file` command. `/help` and `/buffer` didn't work, but those might be related to the `fzf` issue rather than `blink.nvim`.

**Reply by joshuarubin** [reply]
@bassamsdata I'm having the same issue where fzf-lua does not show for certain tools like `/buffers`. I've found that this is related to a bug in codecompanion where it is trying to set the text of a buffer with an invalid (out of range) end column. I've made a PR (https://github.com/olimorris/codecompanion.nvim/pull/624) to fix this, or at least demonstrate the issue.
